Src/VimCore/ExternalEdit.fs

```diff
@@ -4,15 +4,32 @@ namespace Vim
 
 type internal ExternalEditMode(_vimBufferData: IVimBufferData) =
     
+    member x.ClearSelection () =
+        let textView = _vimBufferData.TextView
+        let start = textView.Selection.Start
+        textView.Selection.Clear()
+        TextViewUtil.MoveCaretToVirtualPoint textView start
+
     interface IMode with 
         member x.VimTextBuffer = _vimBufferData.VimTextBuffer
         member x.ModeKind = ModeKind.ExternalEdit
         member x.CommandNames = Seq.empty
-        member x.CanProcess keyInput = keyInput = KeyInputUtil.EscapeKey
+        member x.CanProcess keyInput =
+            keyInput = KeyInputUtil.EscapeKey 
+            || keyInput = KeyInputUtil.CharWithControlToKeyInput 'c'
         member x.Process keyInputData = 
-            if keyInputData.KeyInput = KeyInputUtil.EscapeKey then
-                ProcessResult.OfModeKind ModeKind.Normal
-            else
+            match keyInputData.KeyInput with

```

Comment:
Rather than hooking mode switch did you consider just putting a more direct API on `ExternalEditMode`? The issue with hooking mode switch is that it's more general purpose and cancelling existing operations is generally implicitly true. For instance you cancel word completion when we transition insert to normal mode. Refactoring is the odd one here in that it prefers not to cancel on change and has to be told to. I've been trying to think of other features which would want to do this and can't think of any. So this seems pretty specific to external edit and wondered if you'd thought about putting the API there.

Reply:
The current separation of concerns prevents any direct communication between VimCore and ExternalEditMode. As a result, VimCore must somehow signal indirectly that a general "cancel operation" action occurred which the inline rename can then notice. That event must originate at the VimBuffer level or higher because we can press `<C-c>` in any mode. But the encapsulation is so severe that modes don't even have access to any VimBuffer APIs.

Internal actions like word completion have the luxury participating directly in the logic of the mode doing the cancelling. But external operations like inline rename can't hook `<C-c>` through any of the normal mechanisms. If a "cancel operation" event is insufficient, we would need something like an `IExternalOperation.Cancel` API.

Reply:
Here are some discussion comments about the new architecture that occurred in the issue this PR is intended to close:

- https://github.com/VsVim/VsVim/issues/2587#issuecomment-499340066
- https://github.com/VsVim/VsVim/issues/2587#issuecomment-500098665

The important point is that `ExternalEditMode` can be completely bypassed, but we **still** need a way to cancel refactoring without resorting to the mouse.

Reply:
Upon reflection, I think a "cancel operation" as a mode argument is reasonable. It means "upon entering the new mode, any in-progress host operation that wishes to participate in the "cancel operation" mechanism should now cancel that operation.